### PR TITLE
CNF-11234: Enable RTE metrics to be scraped securely by Prometheus

### PR DESCRIPTION
**Pull request by rbaturov**
This is a follow-up PR for [Enable NROP metrics to be to scraped securely by Prometheus ](https://github.com/openshift-kni/numaresources-operator/pull/1106).

This PR encompasses and integrates all the needed infrastructure to enable secured communication for scraping metrics by Prometheus, now for **RTE**.

A follow-up PR will be posted for e2e tests later.

 To validate that this PR is functioning correctly, please follow these steps:
1. build image of the operator (`make docker-build docker-push`)
2. run: `make deploy` 
3. Attach to one of the prometheus pods `oc exec -it prometheus-k8s-0 -n openshift-monitoring /bin/bash`
4. run:
  ```
 curl -v \
  --cacert /etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt \
  -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" \
  https://numaresources-rte-metrics-service.numaresources.svc:2112/metrics
```

### CODE REVIEW

File: config/rbac/role.yaml [hunk 1]
```diff
@@ -26,6 +26,18 @@ rules:
...
+  resources:
+  - serviceaccounts
+  verbs:
+  - '*'
+- apiGroups:
+  - ""
+  resources:
+  - services
+  verbs:
+  - '*'
```
Comment (ffromani) [inline config/rbac/role.yaml]: ok, this is a role, hence it is ONLY granted in the NROP namespace, right? Stll this is very broad, we need to narrow down AND document in the commit message why do we need these permissions

File: controllers/numaresourcesoperator_controller.go [hunk 2]
```diff
@@ -581,6 +581,31 @@ func (r *NUMAResourcesOperatorReconciler) syncNUMAResourcesOperatorResources(ctx
 			return nil, fmt.Errorf("failed to apply (%s) %s/%s: %w", objState.Desired.GetObjectKind().GroupVersionKind(), objState.Desired.GetNamespace(), objState.Desired.GetName(), err)
 		}
 	}
+
+	for _, obj := range r.RTEMetricsManifests.ToObjects() {
+		// Check if the object already exists
+		existingObj := obj.DeepCopyObject().(client.Object)
```
Comment (ffromani) [inline controllers/numaresourcesoperator_controller.go:587]: why do we need the cast here?

File: controllers/numaresourcesoperator_controller_test.go [hunk 3]
```diff
@@ -23,6 +23,7 @@ import (
 
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
+	rtemetricsmanifests "github.com/openshift-kni/numaresources-operator/pkg/metrics/manifests/monitor"
```
Comment (ffromani) [inline controllers/numaresourcesoperator_controller_test.go:26]: misplaced import

**Review by ffromani — COMMENTED** [review]
I'm not sure we need a separate manifest set. I'. mot saying I'm against it but I'd like to see if/how we can merge it with the existing manifest set

File: controllers/numaresourcesoperator_controller.go (hunk 2 above)
Comment (rbaturov) [reply controllers/numaresourcesoperator_controller.go:587]: If I'll remove the casting:
I'll get this error when using `r.Client.Get`:
`cannot use existingObj (variable of type runtime.Object) as client.Object value in argument to r.Client.Get: runtime.Object does not implement client.Object (missing method GetAnnotations)`

**Comment by rbaturov** [comment]
> I'm not sure we need a separate manifest set. I'. mot saying I'm against it but I'd like to see if/how we can merge it with the existing manifest set

I agree on this one.
But I coudln't figure out a more appropriate place to put this manifest.

File: config/rbac/role.yaml (hunk 1 above)
Comment (rbaturov) [reply config/rbac/role.yaml]: I agree, you are right. we want in only in the NROP namespace.

Comment (rbaturov) [reply config/rbac/role.yaml]: @ffromani 
I've tried to create a separate role (not clusterRole) and a RoleBinding.
But adding the label 
//+kubebuilder:rbac:groups="",resources=services,verbs=*
followed by `make manifests` appends its in the clusterRole.
Not sure how we can avoide this and append this in the dedicated role.

File: controllers/numaresourcesoperator_controller.go (hunk 2 above)
Comment (ffromani) [reply controllers/numaresourcesoperator_controller.go:587]: and why do we use `DeepCopyObject()` and not `DeepCopy()` ?

Comment (rbaturov) [reply controllers/numaresourcesoperator_controller.go:587]: @ffromani This because the client.Object implements DeepCopyObject() and not  DeepCopy()

**Comment by rbaturov** [comment]
> /hold
> 
> need to check RBAC rules again

Regarding this, it is correct that now I have granted permission to the operator to manage service objects across the entire cluster and not narrowed it down to NROP namespace.
As I said before, I'm not sure how we can create a role not a clusterrole for this matter.

**Comment by ffromani** [comment]
> /retest-required

the first retest it's a freebie, but past that we need to check why tests are failing and only when we are confident it's actually an infra issue we should retest

**Comment by rbaturov** [comment]
> > /retest-required
> 
> the first retest it's a freebie, but past that we need to check why tests are failing and only when we are confident it's actually an infra issue we should retest

I saw there was a timeout. I'm not sure this is the infra indeed, I wanted to give it a try before deep diving into that

**Review by ffromani — COMMENTED** [review]
/lgtm

we can narrow down permissions later on

File: controllers/numaresourcesoperator_controller.go [hunk 4]
```diff
@@ -581,6 +583,43 @@ func (r *NUMAResourcesOperatorReconciler) syncNUMAResourcesOperatorResources(ctx
...
+		if apierrors.IsNotFound(err) {
+			err := controllerutil.SetControllerReference(instance, obj, r.Scheme)
+			if err != nil {
+				return nil, fmt.Errorf("failed to set controller reference to %s %s: %w", obj.GetNamespace(), obj.GetName(), err)
+			}
+			err = r.Client.Create(ctx, obj)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create %s/%s: %w", obj.GetNamespace(), obj.GetName(), err)
+			}
+		} else {
```
Comment (Tal-or) [inline controllers/numaresourcesoperator_controller.go:603]: The only change I would suggest here to make the logic more clear is to encapsulate the `else` in:
`UpdateIfNeeded()`  function
The function won't call `Update()` if the objects are actually equals

```
err := r.Client.Get(ctx, client.ObjectKeyFromObject(obj), existingObj)
		if err != nil && !apierrors.IsNotFound(err) {
			return nil, fmt.Errorf("failed to get %s/%s: %w", obj.GetNamespace(), obj.GetName(), err)
		}
		if apierrors.IsNotFound(err) {
			err := controllerutil.SetControllerReference(instance, obj, r.Scheme)
			if err != nil {
				return nil, fmt.Errorf("failed to set controller reference to %s %s: %w", obj.GetNamespace(), obj.GetName(), err)
			}
			err = r.Client.Create(ctx, obj)
			if err != nil {
				return nil, fmt.Errorf("failed to create %s/%s: %w", obj.GetNamespace(), obj.GetName(), err)
			}
		} else {
                           if err := UpdateIfNeeded(ctx, obj, existingObj); err != nil {
                             ....
                           }
}

func UpdateIfNeeded()...
```

**Comment by Tal-or** [comment]
/lgtm
Good enough for now. Thanks!

**Pull request merged**